src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -19,348 +19,161 @@ public static partial class JsonSerializer
         internal static readonly byte[] s_valuesPropertyName
             = new byte[] { (byte)'$', (byte)'v', (byte)'a', (byte)'l', (byte)'u', (byte)'e', (byte)'s' };
 
-        /// <summary>
-        /// Returns true if successful, false is the reader ran out of buffer.
-        /// Sets state.Current.ReturnValue to the reference target for $ref cases;
-        /// Sets state.Current.ReturnValue to a new instance for $id cases.
-        /// </summary>
-        internal static bool ResolveMetadataForJsonObject<T>(
-            ref Utf8JsonReader reader,
-            ref ReadStack state,
-            JsonSerializerOptions options)
+        internal static bool TryReadMetadata(JsonConverter converter, ref Utf8JsonReader reader, ref ReadStack state)
         {
-            JsonConverter converter = state.Current.JsonTypeInfo.PropertyInfoForTypeInfo.ConverterBase;
-
-            if (state.Current.ObjectState < StackFrameObjectState.ReadAheadNameOrEndObject)
-            {
-                // Read the first metadata property name.
-                if (!TryReadAheadMetadataAndSetState(ref reader, ref state, StackFrameObjectState.ReadNameOrEndObject))
-                {
-                    return false;
-                }
-            }
+            Debug.Assert(state.Current.ObjectState == StackFrameObjectState.StartToken);
+            Debug.Assert(state.CanContainMetadata);
 
-            if (state.Current.ObjectState == StackFrameObjectState.ReadNameOrEndObject)
+            while (true)
```

Comment:
This is the key change of the refactoring: rather than using distinct states for each metadata property and forcing a strict ordering for properties, this approach loops over metadata properties (validating any ordering constraints where necessary) much like the object or dictionary converters loop over regular properties.